### DISCUSSION TITLE
Would it be possible to use the eModbusBridge over LoRa?

### DISCUSSION THREAD

**Discussion by HavKrg** [eModbus/eModbus#163] [Q&A]
I am currently using a mess of a program I've written that sends a request string over LoRa, which is then parsed into a modbus-request executed over WiFi, the reply-data is then parsed into a reply-string that is back to the original LoRa Server and parsed into modbus-data which can then be queried with Modbus over WiFi.

Would it be possible to replace this functionality with the eModbusBridge, it would look something like this:

PLC requesting data---->WiFi---->ModbusBridge---->LoRa---->ModbusBridge---->WiFi---->PLC  that returns data

And then return back the same way it came.

Would something like this be possible?

**Comment by Miq1** [comment]
At first thought I do not see anything preventing the use as you described. I do know nothing in regards to LoRa, though.
You may not need a setup that complicated, if it were possible to transmit the Modbus requests and responses via LoRa. Is there a separate communications module connected to SPI, I2C or UART, or is the LoRa functionality integrated into the MCU? Does LoRa have any communications overhead (addressing, checksums etc.) or is it transmitting whatever you will give it?

**Reply by HavKrg** [reply]
Thank you for the quick response.

It uses a separate chip, I am using a TTGO-version as seen here:

https://github.com/LilyGO/TTGO-LoRa32-V2.1

I am currently using the RadioHead-library which has addressing, routing, and even MESH-capabilities, but at the end of the day all LoRa does is take a byte-array and transmit it over a radio connection.

https://www.airspayce.com/mikem/arduino/RadioHead/index.html



I could remove a lot of the parsing in my current setup, but then I would have to change the code at all locations if I wanted to change what data is requested. With the parsing all I have to do is change a string at the "main site"

**Reply by Miq1** [reply]
I see. So this is basically an ESP32 with the LoRa circuitry on board connected via SPI and driven by the RadioHead library.
The PLC on the far side: is it a single Modbus TCP server or several?
It looks like the leanest solution would be to write some Modbus TCP over LoRa tunnel. This could be done similarly to the TCP stub I have written for the library's test sketch, that is providing the usual TCP connect() etc. calls to the application, but does a completely different handling in the background. In your case it would handle the LoRa communications. With such a stub on both ends the communication would look to the PLCs like regular Modbus TCP.

**Reply by HavKrg** [reply]
Sorry for the late response. 

For my use-case there is one main location with 1 TCP are two remote locations each containing 1 TCP server. 

<a href="https://ibb.co/f176VXT"><img src="https://i.ibb.co/f176VXT/Lora-Setup.png" alt="Lora-Setup" border="0"></a>

I am sure there are scenarios where there are multiple PLCs at each remote location, but in those scenarios one could probably just have those communicate with each other, and have one of them communicate with the ESP32.

**Reply by Miq1** [reply]
I see. The LoRa communication on the "central" ESP32 will know which path to use for different IP addresses?
The TCP stub solution I described above would probably work for you. You would have a ModbusBridgeTCP running on all three ESP32s, the central one using a TCP client running the LoRa stub, while the other two will have the stub on their servers' sides of the bridge.
You probably will have to use generous timeout values I reckon, to compensate for slower LoRa communications.

**Reply by HavKrg** [reply]
Yes, that sounds correct to me.  

The central PLC sends out a request. Its local ESP32 will send it either to "remote location A" or "remote location B".  

I guess if the PLC at "remote location A" has an IP of 10.0.0.150, and the ESP32 at that location has a LoRa address of 5, any Modbus request from the PLC at the "central location" to IP address 10.0.0.150 would be routed to LoRa address 5, and then bridged to its PLC. 

Yes, the timeout would have to be quite generous depending on the distance. Several seconds at least.

**Reply by Miq1** [reply]
Exactly. on the "client bridge" ESP32 the LoRa stub would need an address translation table. Ideally the IP address would be part of the LoRa payload, so the receiving "server bridges" will know the IP to address without the need for another address translation table.
